<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Дерево-Решений" data-toc-modified-id="Дерево-Решений-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span><strong>Дерево Решений</strong></a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.0.2"><span class="toc-item-num">2.0.2&nbsp;&nbsp;</span><strong>Случайный лес</strong></a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.0.3"><span class="toc-item-num">2.0.3&nbsp;&nbsp;</span><strong>Логистическая регрессия</strong></a></span></li></ul></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Балансирование-классов" data-toc-modified-id="Балансирование-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Балансирование классов</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span><strong>Дерево решений</strong></a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span><strong>Случайный лес</strong></a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span><strong>Логистическая регрессия</strong></a></span></li></ul></li><li><span><a href="#Увеличение-объектов-редкого-класса" data-toc-modified-id="Увеличение-объектов-редкого-класса-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Увеличение объектов редкого класса</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span><strong>Дерево решений</strong></a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span><strong>Случайный лес</strong></a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span><strong>Логистическая регрессия</strong></a></span></li></ul></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Уменьшение выборки</a></span><ul class="toc-item"><li><span><a href="#Дерево-Решений" data-toc-modified-id="Дерево-Решений-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span><strong>Дерево Решений</strong></a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span><strong>Случайный лес</strong></a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.3.3"><span class="toc-item-num">3.3.3&nbsp;&nbsp;</span><strong>Логистическая регрессия</strong></a></span></li></ul></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Итоговый-вывод" data-toc-modified-id="Итоговый-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Итоговый вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Нужно построить модель с предельно большим значением *F1*-меры. Нужно довести метрику до 0.59 и проверить *F1*-меру на тестовой выборке самостоятельно.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Подготовка данных

In [1]:
#подключение необходимых библиотек

import pandas as pd
import matplotlib.pyplot as plt

from sklearn.preprocessing import(StandardScaler, OrdinalEncoder)
from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split
from sklearn.metrics import (accuracy_score,
                             f1_score, 
                             recall_score, 
                             precision_score, 
                             roc_auc_score, roc_curve, 
                             confusion_matrix)

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.dummy import DummyClassifier

In [2]:
#откроытие данных
data = pd.read_csv("/datasets/Churn.csv")

In [3]:
#вывод первых строк
print(data.head())

   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         790

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


*    RowNumber — индекс строки в данных
*    CustomerId — уникальный идентификатор клиента
*    Surname — фамилия
*    CreditScore — кредитный рейтинг
*    Geography — страна проживания
*    Gender — пол
*    Age — возраст
*    Tenure — сколько лет человек является клиентом банка (**наблюдаются пропуски данных**)
*    Balance — баланс на счёте
*    NumOfProducts — количество продуктов банка, используемых клиентом
*    HasCrCard — наличие кредитной карты
*    IsActiveMember — активность клиента
*    EstimatedSalary — предполагаемая зарплата

**Exited - целевой признак.**

In [5]:
#просмотр дубликатов 
data.duplicated().sum()

0

В стоблце **Tenure** обнаружены пропуски. Заполним их медианным значением.

In [6]:
#заполнение пропусков минимальным значением
data['Tenure'] = data['Tenure'].fillna(data['Tenure'].median())

Так же приведем названия столбцов к "змеиному" стилю.

In [7]:
#переименование столбцов
data.columns = ['row_number', 'customer_id', 'surname', 'credit_score',\
                'geography', 'gender', 'age', 'tenure', 'balance',\
                'num_of_products', 'has_cr_crd', 'is_active_member', 'estimated_salary','exited'\
               ]

In [8]:
#проверка названий столбцов и данных в них
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   row_number        10000 non-null  int64  
 1   customer_id       10000 non-null  int64  
 2   surname           10000 non-null  object 
 3   credit_score      10000 non-null  int64  
 4   geography         10000 non-null  object 
 5   gender            10000 non-null  object 
 6   age               10000 non-null  int64  
 7   tenure            10000 non-null  float64
 8   balance           10000 non-null  float64
 9   num_of_products   10000 non-null  int64  
 10  has_cr_crd        10000 non-null  int64  
 11  is_active_member  10000 non-null  int64  
 12  estimated_salary  10000 non-null  float64
 13  exited            10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


**Вывод:**

В наборе данных есть ненужные столбцы для данного исследования (**row_number, customer_id, surname**), а так же столбцы для более тщательного изучения. **Geography, Gender** преобразуем методом OHE.

## Исследование задачи

Начнем с подготовки данных, а именно: удалим ненужные столбцы(**row_number, customer_id, surname**) и сохраним получившийся датафрейм.

In [9]:
#удаление столбцов
data = data.drop(['row_number', 'customer_id', 'surname'], axis=1)

#просмотр готового варианта
data.head()

credit_score geography  gender  age  tenure    balance  num_of_products  \
0           619    France  Female   42     2.0       0.00                1   
1           608     Spain  Female   41     1.0   83807.86                1   
2           502    France  Female   42     8.0  159660.80                3   
3           699    France  Female   39     1.0       0.00                2   
4           850     Spain  Female   43     2.0  125510.82                1   

   has_cr_crd  is_active_member  estimated_salary  exited  
0           1                 1         101348.88       1  
1           0                 1         112542.58       0  
2           1                 0         113931.57       1  
3           0                 0          93826.63       0  
4           1                 1          79084.10       0

In [10]:
#просмотр уникальных значений для дальнейшего преобразования OHE
data['geography'].value_counts()

France     5014
Germany    2509
Spain      2477
Name: geography, dtype: int64

In [11]:
#просмотр уникальных значений для дальнейшего преобразования OHE
data['gender'].value_counts()

Male      5457
Female    4543
Name: gender, dtype: int64

In [12]:
#информация о данных до преобразования OHE
data.shape

(10000, 11)

У нас получился датафрейм **data** с 10000 строк и 11 стобцами, куда входят несколько категориальных столбцов(**geography с 3 разновидностями и gender с 2 разновидностями**). Теперь стоит подготовить данные методом OHE во избежании дамми-ловушек для дальнейшего использования в Логистической регрессии (**data_ohe**), а так же для решающего дерева и случайного леса используем порядковое кодирование(**data_ordinal**).

In [13]:
#OHE
data_ohe = pd.get_dummies(data, drop_first=True)

#вывод на экран
data_ohe.head()

credit_score  age  tenure    balance  num_of_products  has_cr_crd  \
0           619   42     2.0       0.00                1           1   
1           608   41     1.0   83807.86                1           0   
2           502   42     8.0  159660.80                3           1   
3           699   39     1.0       0.00                2           0   
4           850   43     2.0  125510.82                1           1   

   is_active_member  estimated_salary  exited  geography_Germany  \
0                 1         101348.88       1                  0   
1                 1         112542.58       0                  0   
2                 0         113931.57       1                  0   
3                 0          93826.63       0                  0   
4                 1          79084.10       0                  0   

   geography_Spain  gender_Male  
0                0            0  
1                1            0  
2                0            0  
3                0            0  
4                1            0

In [14]:
data_ohe.shape

(10000, 12)

После преобразования методом OHE у нас стало 12 столбцов, при этом из **geography** получилось 2 класса вместо 3, а из **gender** 1 вместо 2.

In [15]:
#порядковое кодирование
encoder = OrdinalEncoder()
data_ordinal = pd.DataFrame(encoder.fit_transform(data), columns=data.columns)

После преобразования можно приступить к созданию валидационной, обучающей и тестовой выборкам. Данные поделим стандартно в соотношении 60-20-20.

In [16]:
#разделение по признакам
target_ohe = data_ohe['exited']
features_ohe = data_ohe.drop('exited', axis=1)

#разделение на выборки OHE
#валидационная выборка (60%)
features_train_ohe, features_valid_ohe, target_train_ohe, target_valid_ohe = train_test_split\
(features_ohe, target_ohe, test_size=0.6, random_state=12345)

#обучающая и тестовая (по 20%)
features_valid_ohe, features_test_ohe, target_valid_ohe, target_test_ohe = train_test_split\
(features_valid_ohe, target_valid_ohe, test_size=0.5, random_state=12345)

In [17]:
#разделение по признакам
target_ordinal = data_ordinal['exited']
features_ordinal = data_ordinal.drop('exited', axis=1)

#разделение на выборки Ordinal
#валидационная выборка (60%)
features_train_ordinal, features_valid_ordinal,\
target_train_ordinal, target_valid_ordinal = train_test_split (features_ordinal, target_ordinal,\
                                                               test_size=0.6, random_state=12345)

#обучающая и тестовая (по 20%)
features_valid_ordinal, features_test_ordinal,\
target_valid_ordinal, target_test_ordinal = train_test_split(features_valid_ordinal,\
                                        target_valid_ordinal, test_size=0.5, random_state=12345)

Теперь приступим к масштабированию количественных признаков. Для этого сохраним их в переменную **numeric** и используем метод **scaler()**.

In [18]:
#сохранение списка количественных признаков
numeric = ['credit_score', 'age', 'tenure', 'balance', 'num_of_products', 'estimated_salary']

#игнорирование предупреждений
import warnings 
warnings.filterwarnings('ignore')

#масштабирование
scaler = StandardScaler()
scaler.fit(features_train_ohe[numeric])
scaler.fit(features_train_ordinal[numeric])

features_train_ohe[numeric] = scaler.transform(features_train_ohe[numeric])
features_valid_ohe[numeric] = scaler.transform(features_valid_ohe[numeric])
features_test_ohe[numeric] = scaler.transform(features_test_ohe[numeric])

features_train_ordinal[numeric] = scaler.transform(features_train_ordinal[numeric])
features_valid_ordinal[numeric] = scaler.transform(features_valid_ordinal[numeric])
features_test_ordinal[numeric] = scaler.transform(features_test_ordinal[numeric])


Теперь переходим к обучению моделей.

#### **Дерево Решений**

In [19]:
#модель дерева решений
best_model = None
best_result = 0
for depth in range (1,20):
    model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_tree.fit(features_train_ordinal, target_train_ordinal)
    predicted_tree = model_tree.predict(features_valid_ordinal)
    result = f1_score(target_valid_ordinal, predicted_tree)
    if result > best_result:
        best_model = model_tree
        best_result = result
                
#Расчет roc_auc для Дерева Решений
probabilities_one_valid_tree = model_tree.predict_proba(features_valid_ordinal)[:, 1]
auc_roc_tree = roc_auc_score(target_valid_ordinal, probabilities_one_valid_tree)
print("F1-мера: ", best_result,"\n",
      "Лучшая модель дерева: ", best_model,'\n',
      "AUC-ROC: ", auc_roc_tree)

F1-мера:  0.5411764705882353 
 Лучшая модель дерева:  DecisionTreeClassifier(max_depth=5, random_state=12345) 
 AUC-ROC:  0.6880133945613048


In [20]:
#построение предсказаний и матрицы ошибок
error_tree = confusion_matrix(target_valid_ordinal, predicted_tree)
error_tree

array([[2017,  397],
       [ 270,  316]])

**Вывод:**

* Из матрицы ошибок: TN (истинно отрицательные) 2017, TP (истинно положительные) 316, FN (ложноотрицательные) 270, FP (ложноположительные) 397. То есть модель хорошо предсказывает TN, но много ошибается в FP.
* Из метрик: F1-мера (качество модели)- 0.5411764705882353, AUC-ROC равна 0.6880133945613048
* Лучшая модель дерева:  DecisionTreeClassifier(max_depth=5, random_state=12345) 

#### **Случайный лес**

In [21]:
#модель леса
best_model = None
best_result = 0
for est in range (1,20):
    for depth in range(1,10):
        for leaf in range(1,10):
            model_forest = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth,\
                                                  min_samples_leaf=leaf)
            model_forest.fit(features_train_ordinal, target_train_ordinal)
            predicted_forest = model_forest.predict(features_valid_ordinal)
            result = f1_score(target_valid_ordinal, predicted_forest)
            if result > best_result:
                best_model = model_forest
                best_result = result

#Расчет roc_auc для Случайного леса
probabilities_one_valid_forest = model_forest.predict_proba(features_valid_ordinal)[:, 1]
auc_roc_forest = roc_auc_score(target_valid_ordinal, probabilities_one_valid_forest)
print("F1-мера: ", best_result,"\n",
      "Лучшая модель леса: ", best_model,'\n',
      "AUC-ROC: ", auc_roc_forest)

F1-мера:  0.566990291262136 
 Лучшая модель леса:  RandomForestClassifier(max_depth=6, min_samples_leaf=3, n_estimators=1,
                       random_state=12345) 
 AUC-ROC:  0.848528634161928


In [22]:
#построение предсказаний и матрицы ошибок
error_forest = confusion_matrix(target_valid_ordinal, predicted_forest)
error_forest

array([[2343,   71],
       [ 335,  251]])

**Вывод:**

* Из матрицы ошибок: TN (истинно отрицательные) 2343, TP (истинно положительные) 251, FN (ложноотрицательные) 335, FP (ложноположительные) 71. То есть модель хорошо предсказывает TN, TP, но дает много FN.
* Из метрик: F1-мера - 0.566990291262136, AUC-ROC равна 0.848528634161928
* Лучшая модель леса:  RandomForestClassifier(max_depth=6, min_samples_leaf=3, n_estimators=1, random_state=12345)

#### **Логистическая регрессия**

In [23]:
#модель Логистической регрессии
model_log = LogisticRegression(random_state=12345, solver='liblinear')
model_log.fit(features_train_ohe, target_train_ohe)
predicted_log = model_log.predict(features_valid_ohe)
result = f1_score(target_valid_ohe, predicted_log)
                    
#Расчет roc_auc для Логистической регрессии
probabilities_one_valid_log = model_log.predict_proba(features_valid_ohe)[:, 1]
auc_roc_log = roc_auc_score(target_valid_ohe, probabilities_one_valid_log)

print("F1-мера: ", result,"\n",
      "AUC-ROC: ", auc_roc_log)

F1-мера:  0.3278301886792453 
 AUC-ROC:  0.7587508588976137


In [24]:
#построение предсказаний и матрицы ошибок
error_log = confusion_matrix(target_valid_ohe, predicted_log)
error_log

array([[2291,  123],
       [ 447,  139]])

**Вывод:**

* Из матрицы ошибок: TN (истинно отрицательные) 2291, TP (истинно положительные) 139, FN (ложноотрицательные) 447, FP (ложноположительные) 123. То есть модель хорошо предсказывает TN, TP, но дает много FN.
* Из метрик: F1-мера - 0.3278301886792453, AUC-ROC равна 0.7587508588976137

In [25]:
data['exited'].value_counts(normalize=True)

0    0.7963
1    0.2037
Name: exited, dtype: float64

Соотношение приблизительно 80% к 20%, что говорит и **дисбалансе.** Посмотрим на матрицы ошибок моделей.

## Борьба с дисбалансом

### Балансирование классов

Придадим большую важность классу 1 с помощью параметра class_weight='balanced' при обучении. После проведем увеличение выборки и уменьшение, после сравним результаты до и после.

#### **Дерево решений**

In [26]:
#модель дерева решений
best_model = None
best_result = 0
for depth in range (1,20):
    model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth, class_weight='balanced')
    model_tree.fit(features_train_ordinal, target_train_ordinal)
    predicted_tree = model_tree.predict(features_valid_ordinal)
    result = f1_score(target_valid_ordinal, predicted_tree)
    if result > best_result:
        best_model = model_tree
        best_result = result
                
#Расчет roc_auc для Дерева Решений
probabilities_one_valid_tree = model_tree.predict_proba(features_valid_ordinal)[:, 1]
auc_roc_tree = roc_auc_score(target_valid_ordinal, probabilities_one_valid_tree)
print("F1-мера: ", best_result,"\n",
      "Лучшая модель дерева: ", best_model,'\n',
      "AUC-ROC: ", auc_roc_tree)

F1-мера:  0.5486725663716814 
 Лучшая модель дерева:  DecisionTreeClassifier(class_weight='balanced', max_depth=5, random_state=12345) 
 AUC-ROC:  0.6888090235853992


**ДО:**

* F1-мера:  0.5411764705882353 
* Лучшая модель дерева:  DecisionTreeClassifier(max_depth=5, random_state=12345) 
* AUC-ROC:  0.6880133945613048

**ПОСЛЕ:**

* F1-мера:  0.5486725663716814 
* Лучшая модель дерева:  DecisionTreeClassifier(class_weight='balanced', max_depth=5, random_state=12345) 
* AUC-ROC:  0.6888090235853992

**Вывод: наблюдается незначительное улучшение метрик.**

#### **Случайный лес**

In [27]:
#модель леса
best_model = None
best_result = 0
for est in range (1,20):
    for depth in range(1,10):
        for leaf in range(1,10):
            model_forest = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth,\
                                                  min_samples_leaf=leaf, class_weight='balanced')
            model_forest.fit(features_train_ordinal, target_train_ordinal)
            predicted_forest = model_forest.predict(features_valid_ordinal)
            result = f1_score(target_valid_ordinal, predicted_forest)
            if result > best_result:
                best_model = model_forest
                best_result = result

#Расчет roc_auc для Случайного леса
probabilities_one_valid_forest = model_forest.predict_proba(features_valid_ordinal)[:, 1]
auc_roc_forest = roc_auc_score(target_valid_ordinal, probabilities_one_valid_forest)
print("F1-мера: ", best_result,"\n",
      "Лучшая модель леса: ", best_model,'\n',
      "AUC-ROC: ", auc_roc_forest)

F1-мера:  0.6118370484242891 
 Лучшая модель леса:  RandomForestClassifier(class_weight='balanced', max_depth=9, min_samples_leaf=2,
                       n_estimators=11, random_state=12345) 
 AUC-ROC:  0.8507635352367164


**ДО:**

* F1-мера:  0.566990291262136 
* Лучшая модель леса:  RandomForestClassifier(max_depth=6, min_samples_leaf=3, n_estimators=1, random_state=12345) 
* AUC-ROC:  0.848528634161928

**ПОСЛЕ:**

* F1-мера:  0.6118370484242891 
* Лучшая модель леса:  RandomForestClassifier(class_weight='balanced', max_depth=9, min_samples_leaf=2, n_estimators=11, random_state=12345) 
* AUC-ROC:  0.8507635352367164

**Вывод: значительное улучшение метрик.**

#### **Логистическая регрессия**

In [28]:
#модель Логистической регрессии
model_log = LogisticRegression(random_state=12345, solver='liblinear', class_weight='balanced')
model_log.fit(features_train_ohe, target_train_ohe)
predicted_log = model_log.predict(features_valid_ohe)
result = f1_score(target_valid_ohe, predicted_log)
                    
#Расчет roc_auc для Логистической регрессии
probabilities_one_valid_log = model_log.predict_proba(features_valid_ohe)[:, 1]
auc_roc_log = roc_auc_score(target_valid_ohe, probabilities_one_valid_log)

print("F1-мера: ", result,"\n",
      "AUC-ROC: ", auc_roc_log)

F1-мера:  0.47757716948165396 
 AUC-ROC:  0.7631103828350547


**ДО:**

* Из метрик: F1-мера - 0.3278301886792453, AUC-ROC равна 0.7587508588976137

**ПОСЛЕ:**

* Из метрик: F1-мера - 0.47757716948165396, AUC-ROC равна 0.7631103828350547

**Вывод: Наблюдается улучшение метрик**

**Итог:** 
* Дерево решений F1-мера: 0.5486725663716814, AUC-ROC: 0.6888090235853992
* Случайный лес  F1-мера: 0.6118370484242891, AUC-ROC: 0.8507635352367164  - **лидер**
* Логистическая регрессия F1-мера - 0.47757716948165396, AUC-ROC 0.7631103828350547

### Увеличение объектов редкого класса

In [30]:
#функция увеличения
def upsample(features, target, repeat=4):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
          
    return shuffle(features_upsampled, target_upsampled, random_state=12345)

In [31]:
#создание признаков для обучения
features_train_ohe_upsampled, target_train_ohe_upsampled = upsample(features_train_ohe,\
                                                                          target_train_ohe)
features_train_ordinal_upsampled, target_train_ordinal_upsampled = upsample(\
                                                       features_train_ordinal, target_train_ordinal)

#### **Дерево решений**

In [32]:
#модель дерева решений
best_model_up = None
best_result_up = 0
for depth in range (1,20):
    model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_tree.fit(features_train_ordinal_upsampled, target_train_ordinal_upsampled)
    predicted_tree_up = model_tree.predict(features_valid_ordinal)
    result_up = f1_score(target_valid_ordinal, predicted_tree_up)
    if result_up > best_result_up:
        best_model_up = model_tree
        best_result_up = result_up
                
#Расчет roc_auc для Дерева Решений
probabilities_one_valid_tree_up = model_tree.predict_proba(features_valid_ordinal)[:, 1]
auc_roc_tree_up = roc_auc_score(target_valid_ordinal, probabilities_one_valid_tree_up)
print("F1-мера: ", best_result_up,"\n",
      "Лучшая модель дерева: ", best_model_up,'\n',
      "AUC-ROC: ", auc_roc_tree_up)

F1-мера:  0.5486725663716814 
 Лучшая модель дерева:  DecisionTreeClassifier(max_depth=5, random_state=12345) 
 AUC-ROC:  0.6910103463584156


**ДО:**

* F1-мера:  0.5486725663716814 
* Лучшая модель дерева:  DecisionTreeClassifier(class_weight='balanced', max_depth=5, random_state=12345) 
* AUC-ROC:  0.6888090235853992

**ПОСЛЕ:**

* F1-мера:  0.5486725663716814 
* Лучшая модель дерева:  DecisionTreeClassifier(max_depth=5, random_state=12345) 
* AUC-ROC:  0.6910103463584156


**Вывод: метрики немного увеличелись.**

#### **Случайный лес**

In [33]:
#модель леса
best_model_up = None
best_result_up = 0
for est in range (1,20):
    for depth in range(1,10):
        for leaf in range(1,10):
            model_forest = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth,\
                                                  min_samples_leaf=leaf)
            model_forest.fit(features_train_ordinal_upsampled, target_train_ordinal_upsampled)
            predicted_forest_up = model_forest.predict(features_valid_ordinal)
            result_up = f1_score(target_valid_ordinal, predicted_forest_up)
            if result_up > best_result_up:
                best_model_up = model_forest
                best_result_up = result_up

#Расчет roc_auc для Случайного леса
probabilities_one_valid_forest_up = model_forest.predict_proba(features_valid_ordinal)[:, 1]
auc_roc_forest_up = roc_auc_score(target_valid_ordinal, probabilities_one_valid_forest_up)
print("F1-мера: ", best_result_up,"\n",
      "Лучшая модель леса: ", best_model_up,'\n',
      "AUC-ROC: ", auc_roc_forest_up)

F1-мера:  0.597163120567376 
 Лучшая модель леса:  RandomForestClassifier(max_depth=9, min_samples_leaf=3, n_estimators=16,
                       random_state=12345) 
 AUC-ROC:  0.8470388179306718


**ДО:**

* F1-мера:  0.6118370484242891 
* Лучшая модель леса:  RandomForestClassifier(class_weight='balanced', max_depth=9, min_samples_leaf=2, n_estimators=11, random_state=12345) 
* AUC-ROC:  0.8507635352367164

**ПОСЛЕ:**

* F1-мера:  0.597163120567376 
*  Лучшая модель леса:  RandomForestClassifier(max_depth=9, min_samples_leaf=3, n_estimators=16, random_state=12345) 
* AUC-ROC:  0.8470388179306718


**Вывод: показатели немного уменьшились. При балансировке значения лучше**

#### **Логистическая регрессия**

In [34]:
#модель Логистической регрессии
model_log = LogisticRegression(random_state=12345, solver='liblinear')
model_log.fit(features_train_ohe_upsampled, target_train_ohe_upsampled)
predicted_log_up = model_log.predict(features_valid_ohe)
result_up = f1_score(target_valid_ohe, predicted_log_up)
                    
#Расчет roc_auc для Логистической регрессии
probabilities_one_valid_log_up = model_log.predict_proba(features_valid_ohe)[:, 1]
auc_roc_log_up = roc_auc_score(target_valid_ohe, probabilities_one_valid_log_up)

print("F1-мера: ", result_up,"\n",
      "AUC-ROC: ", auc_roc_log_up)

F1-мера:  0.4791901012373454 
 AUC-ROC:  0.7632963005901297


**ДО:**

F1-мера - 0.47757716948165396, AUC-ROC: 0.7631103828350547

**ПОСЛЕ:**

F1-мера:  0.4791901012373454, AUC-ROC: 0.7632963005901297


**Вывод: показатели немного увеличелись**

**Итог:**
* Дерево Решений F1-мера: 0.5486725663716814, AUC-ROC: 0.6910103463584156 
* Случайный лес  F1-мера: 0.597163120567376 , AUC-ROC: 0.8470388179306718 **лидер по F1** 
* Логистическая регрессия F1-мера: 0.4791901012373454, AUC-ROC: 0.7632963005901297

Общий итог: показатели снизиль у Случайного леса, у остальных моделей незначительно поднялись по сравнению с показателями после балансировки. Даже при таком раскладе лучшей моделью снова является Случайный лес.

### Уменьшение выборки

In [59]:
#функция уменьшения
def downsample(features, target, fraction = 0.2):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones]
    )
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones]
    )
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

In [60]:
#создание признаков для обучения
features_train_ohe_downsampled, target_train_ohe_downsampled = downsample(features_train_ohe,\
                                                                          target_train_ohe)
features_train_ordinal_downsampled, target_train_ordinal_downsampled = downsample(\
                                                       features_train_ordinal, target_train_ordinal)

#### **Дерево Решений**

In [61]:
#модель дерева решений
best_model_down = None
best_result_down = 0
for depth in range (1,20):
    model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_tree.fit(features_train_ordinal_downsampled, target_train_ordinal_downsampled)
    predicted_tree_down = model_tree.predict(features_valid_ordinal)
    result_down = f1_score(target_valid_ordinal, predicted_tree_down)
    if result_down > best_result_down:
        best_model_down = model_tree
        best_result_down = result_down
                
#Расчет roc_auc для Дерева Решений
probabilities_one_valid_tree_down = model_tree.predict_proba(features_valid_ordinal)[:, 1]
auc_roc_tree_down = roc_auc_score(target_valid_ordinal, probabilities_one_valid_tree_down)
print("F1-мера: ", best_result_down,"\n",
      "Лучшая модель дерева: ", best_model_down,'\n',
      "AUC-ROC: ", auc_roc_tree_down)

F1-мера:  0.5481012658227848 
 Лучшая модель дерева:  DecisionTreeClassifier(max_depth=5, random_state=12345) 
 AUC-ROC:  0.6882781329615921


**ДО(баланс):**

* F1-мера: 0.5486725663716814
* Лучшая модель дерева: DecisionTreeClassifier(class_weight='balanced', max_depth=5, random_state=12345)
* AUC-ROC: 0.6888090235853992

**ПОСЛЕ:**

* F1-мера:  0.5481012658227848 
* Лучшая модель дерева:  DecisionTreeClassifier(max_depth=5, random_state=12345) 
* AUC-ROC:  0.6882781329615921

**Вывод: по сравнению с балансировкой показатели стали чуть меньше.**

#### **Случайный лес**

In [62]:
#модель леса
best_model_down = None
best_result_down = 0
for est in range (1,20):
    for depth in range(1,10):
        for leaf in range(1,10):
            model_forest = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth,\
                                                  min_samples_leaf=leaf)
            model_forest.fit(features_train_ordinal_downsampled, target_train_ordinal_downsampled)
            predicted_forest_down = model_forest.predict(features_valid_ordinal)
            result_down = f1_score(target_valid_ordinal, predicted_forest_down)
            if result_down > best_result_down:
                best_model_down = model_forest
                best_result_down = result_down

#Расчет roc_auc для Случайного леса
probabilities_one_valid_forest_down = model_forest.predict_proba(features_valid_ordinal)[:, 1]
auc_roc_forest_down = roc_auc_score(target_valid_ordinal, probabilities_one_valid_forest_down)
print("F1-мера: ", best_result_down,"\n",
      "Лучшая модель леса: ", best_model_down,'\n',
      "AUC-ROC: ", auc_roc_forest_down)

F1-мера:  0.532249322493225 
 Лучшая модель леса:  RandomForestClassifier(max_depth=8, min_samples_leaf=3, n_estimators=19,
                       random_state=12345) 
 AUC-ROC:  0.833867640696619


**ДО (баланс):**

* F1-мера:  0.6118370484242891 
* Лучшая модель леса:  RandomForestClassifier(class_weight='balanced', max_depth=9, min_samples_leaf=2, n_estimators=11, random_state=12345) 
* AUC-ROC:  0.8507635352367164

**ПОСЛЕ:**

* F1-мера:  0.532249322493225 
* Лучшая модель леса:  RandomForestClassifier(max_depth=8, min_samples_leaf=3, n_estimators=19, random_state=12345) 
* AUC-ROC:  0.833867640696619

**Вывод: по сравнению с балансировкой показатели стали меньше.**

#### **Логистическая регрессия**

In [63]:
#модель Логистической регрессии
model_log = LogisticRegression(random_state=12345, solver='liblinear')
model_log.fit(features_train_ohe_downsampled, target_train_ohe_downsampled)
predicted_log_down = model_log.predict(features_valid_ohe)
result_down = f1_score(target_valid_ohe, predicted_log_down)
                    
#Расчет roc_auc для Логистической регрессии
probabilities_one_valid_log_down = model_log.predict_proba(features_valid_ohe)[:, 1]
auc_roc_log_down = roc_auc_score(target_valid_ohe, probabilities_one_valid_log_down)

print("F1-мера: ", result_down,"\n",
      "AUC-ROC: ", auc_roc_log_down)

F1-мера:  0.4629905613512171 
 AUC-ROC:  0.7655379173252727


**ДО (баланс):**

F1-мера - 0.47757716948165396, AUC-ROC: 0.7631103828350547

**ПОСЛЕ:**

F1-мера:  0.4629905613512171, AUC-ROC:  0.7655379173252727

**Вывод: Незначительное уменьшение показателей по сравнению с балансировкой**

**Итог:** 

* Дерево решений F1-мера: 0.5481012658227848, AUC-ROC: 0.68827813296159212
* Случайный лес  F1-мера: 0.532249322493225, AUC-ROC: 0.833867640696619  **лидер**
* Логистическая регрессия F1-мера: 0.4629905613512171, AUC-ROC: 0.7655379173252727

В результате лучшие показатели у модели Случайного леса при балансировке. **RandomForestClassifier(class_weight='balanced', random_state=12345, n_estimators=11, max_depth=9, min_samples_leaf=2).**

In [64]:
#модель леса
model_forest = RandomForestClassifier(class_weight='balanced', random_state=12345, n_estimators=11,\
                                      max_depth=9, min_samples_leaf=2)
model_forest.fit(features_train_ordinal, target_train_ordinal)
predicted_forest = model_forest.predict(features_valid_ordinal)
result_forest = f1_score(target_valid_ordinal, predicted_forest)

#Расчет roc_auc для Случайного леса
probabilities_one_valid_forest = model_forest.predict_proba(features_valid_ordinal)[:, 1]
auc_roc_forest_best = roc_auc_score(target_valid_ordinal, probabilities_one_valid_forest)
print("F1-мера: ", result_forest,"\n",
      "AUC-ROC: ", auc_roc_forest_best)

F1-мера:  0.6118370484242891 
 AUC-ROC:  0.8457426954822693


Показатели высокие. Теперь протестируем эту модель.

## Тестирование модели

Проверим остальные метрики.

In [65]:
predicted_forest_final = model_forest.predict(features_test_ordinal)

#метрики полноты, точности, F1 меры
rec_forest_final = recall_score(target_test_ordinal, predicted_forest_final)    
pre_forest_final = precision_score(target_test_ordinal, predicted_forest_final)
f1_forest_final = f1_score(target_test_ordinal, predicted_forest_final)

#Расчет roc_auc
probabilities_forest_final = model_forest.predict_proba(features_test_ordinal)[:, 1]
auc_roc_forest_final = roc_auc_score(target_test_ordinal, probabilities_forest_final)

#вывод на экран значений
print("Полнота: ", rec_forest_final,"\n",\
      "Точность:", pre_forest_final, "\n",\
      "F1-мера:", f1_forest_final, "\n",\
      "AUC-ROC:", auc_roc_forest_final)

Полнота:  0.6177924217462932 
 Точность: 0.5498533724340176 
 F1-мера: 0.5818463925523663 
 AUC-ROC: 0.8413708021267411


In [66]:
#константая модель
target_pred_constant = pd.Series(0, index=target_test_ordinal)
target_pred_constant.value_counts()

0    3000
dtype: int64

In [67]:
#Сравним показатели
print('accuracy_score константой модели:', accuracy_score(target_valid_ordinal, target_pred_constant))
print('accuracy_score финальной модели:', accuracy_score(target_test_ordinal, predicted_forest_final))
print('AUC-ROC константой модели:', roc_auc_score(target_valid_ordinal, target_pred_constant))
print('AUC-ROC финальной модели:', roc_auc_score(target_test_ordinal, predicted_forest_final))

accuracy_score константой модели: 0.8046666666666666
accuracy_score финальной модели: 0.8203333333333334
AUC-ROC константой модели: 0.5
AUC-ROC финальной модели: 0.7447507867193647


Результаты финальной модели выше чем константной. Модель адекватна.

## Итоговый вывод

Изначально датасет был разбалансирован в соотношении 80/20 в пользу отрицательных результатов. На основе разбалансировки были достигнуты следующие результаты:

* **Дерево решений:**
    -   Из матрицы ошибок: TN (истинно отрицательные) 2017, TP (истинно положительные) 316, FN (ложноотрицательные) 270, FP (ложноположительные) 397. То есть модель хорошо предсказывает TN, но много ошибается в FP.
    -    Из метрик: F1-мера (качество модели)- 0.5411764705882353, AUC-ROC равна 0.6880133945613048
    -    Лучшая модель дерева: DecisionTreeClassifier(max_depth=5, random_state=12345)

* **Случайный лес:**
    -    Из матрицы ошибок: TN (истинно отрицательные) 2343, TP (истинно положительные) 251, FN (ложноотрицательные) 335, FP (ложноположительные) 71. То есть модель хорошо предсказывает TN, TP, но дает много FN.
    -    Из метрик: F1-мера - 0.566990291262136, AUC-ROC равна 0.848528634161928
    -    Лучшая модель леса: RandomForestClassifier(max_depth=6, min_samples_leaf=3, n_estimators=1, random_state=12345)

* **Логистическая регрессия:**
    -    Из матрицы ошибок: TN (истинно отрицательные) 2291, TP (истинно положительные) 139, FN (ложноотрицательные) 447, FP (ложноположительные) 123. То есть модель хорошо предсказывает TN, TP, но дает много FN.
    -    Из метрик: F1-мера - 0.3278301886792453, AUC-ROC равна 0.7587508588976137




После этого были использованы методы борьбы с дисбалансом:**балансирование классов, увеличение объектов редкого класса, уменьшение выборки**. Наулучшей моделью себя показала **Случайный лес** при методе **балансировки классов** со следующими показателями: F1-мера - 0.6118370484242891, AUC-ROC - 0.8457426954822693.

В дальнейшем на тестовой выборке была использована модель **RandomForestClassifier(class_weight='balanced', random_state=12345, n_estimators=11, max_depth=9, min_samples_leaf=2)**, которая дала следующие результаты: 
* Полнота:  0.6177924217462932 
* Точность: 0.5498533724340176 
* F1-мера: 0.5818463925523663 
* AUC-ROC: 0.8413708021267411


При сравнении с константной моделью получились следующие результаты: 

* accuracy_score константой модели: 0.8046666666666666
* accuracy_score финальной модели: 0.8203333333333334
* AUC-ROC константой модели: 0.5
* AUC-ROC финальной модели: 0.7447507867193647

Финальная модель продемонстрировала более высокие показатели.